tests/gstreamer_mqtt/unittest_ntp_util_mock.cc

```diff
@@ -0,0 +1,317 @@
+using ::testing::Return;
+using ::testing::SetArgPointee;
+using ::testing::Assign;
+using ::testing::SetErrnoAndReturn;
+
+const uint64_t NTPUTIL_TIMESTAMP_DELTA = 2208988800ULL;
+
+/**
+ * @brief Interface for NTP util mock class
+ */
+class INtpUtil {
+public:
+  /**
+   * @brief Destroy the INtpUtil object
+   */
+  virtual ~INtpUtil () {};
+  virtual struct hostent *gethostbyname (const char *name) = 0;
+  virtual int connect (int sockfd, const struct sockaddr *addr,
+                  socklen_t addrlen) = 0;
+  virtual ssize_t write (int fd, const void *buf, size_t count) = 0;
+  virtual ssize_t read (int fd, void *buf, size_t count) = 0;
+  virtual uint32_t  _convert_to_host_byte_order (uint32_t netlong) = 0;
+};
+
+/**
+ * @brief Mock class for testing ntp util
+ */
+class NtpUtilMock : public INtpUtil {
+public:
+  MOCK_METHOD (struct hostent *, gethostbyname, (const char *name));
+  MOCK_METHOD (int, connect, (int sockfd, const struct sockaddr *addr,
+                  socklen_t addrlen));
+  MOCK_METHOD (ssize_t, write, (int fd, const void *buf, size_t count));
+  MOCK_METHOD (ssize_t, read, (int fd, void *buf, size_t count));
+  MOCK_METHOD (uint32_t, _convert_to_host_byte_order, (uint32_t netlong));
+};
+NtpUtilMock *mockInstance = nullptr;
+
+/**
+ * @brief Mocking function for gethostbyname
+ */
+struct hostent *gethostbyname (const char *name)
+{
+  return mockInstance->gethostbyname (name);
+}
+
+/**
+ * @brief Mocking function for gethostbyname
+ */
+int connect (int sockfd, const struct sockaddr *addr,
+                  socklen_t addrlen)
+{
+  return mockInstance->connect (sockfd, addr, addrlen);
+}
+
+/** @note To avoid redundant declaration in the test */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
+/**
+ * @brief Mocking function for write.
+ */
+ssize_t write (int fd, const void *buf, size_t count)
+{
+  return mockInstance->write (fd, buf, count);
+}
+
+/**
+ * @brief Mocking function for read.
+ */
+ssize_t read (int fd, void *buf, size_t count)
+{
+  return mockInstance->read (fd, buf, count);
+}
+#pragma GCC diagnostic pop
+
+/**
+ * @brief Mocking function for _convert_to_host_byte_order.
+ */
+uint32_t _convert_to_host_byte_order (uint32_t netlong)
+{
+  return mockInstance->_convert_to_host_byte_order (netlong);
+}
+
+/**
+ * @brief Initialize hostent.
+ */
+static void _hostent_init (struct hostent *host)
+{
+  host->h_name = g_strdup ("github.com");
+  host->h_aliases = (char **) calloc (1, sizeof (char *));
+  host->h_aliases[0] = g_strdup ("www.github.com");
+  host->h_addrtype = AF_INET;
+  host->h_length = 4;
+  host->h_addr_list = (char **) calloc (1, sizeof (char *));
+  host->h_addr_list[0] = g_strdup ("52.78.231.108");
+}
+
+/**
+ * @brief Free hostent.
+ */
+static void _hostent_free (struct hostent *host)
+{
+  g_free (host->h_name);
+  g_free (host->h_aliases[0]);
+  g_free (host->h_aliases);
+  g_free (host->h_addr_list[0]);
+  g_free (host->h_addr_list);
+}
```

Comment:
I found out that every test case calls _hostent_init() before test and _hostent_free() after test.
What about applying the Test Fixture of Google Test instead of calling both of them?

For example, `NtpUtilMockTest` class for Test Fixture can be defined as below.

```cpp
clsss NtpUtilMockTest : public::testing::Test {
  ...
  
protected:
  struct hostent host;
  ...
  
public:
  void SetUp() override
  {
    host.h_name = g_strdup ("github.com");
    host.h_aliases = (char **) calloc (1, sizeof (char *));
    host.h_aliases[0] = g_strdup ("www.github.com");
    host.h_addrtype = AF_INET;
    host.h_length = 4;
    host.h_addr_list = (char **) calloc (1, sizeof (char *));
    host.h_addr_list[0] = g_strdup ("52.78.231.108");
  }
  
  void TearDown() override
  {
    g_free (host.h_name);
    g_free (host.h_aliases[0]);
    g_free (host.h_aliases);
    g_free (host.h_addr_list[0]);
    g_free (host.h_addr_list);
  }
};
```

Reply:
Right. This method looks better. I applied test fixture. Thanks!